### PR TITLE
Catchpoints:  Enrich catchpoint generation and status with KV metadata

### PR DESCRIPTION
**Pull request by michaeldiamant**
Enriches catchpoints with the following KV metadata mirroring _accounts_:
* Catchpoint creation (writes) - Add total KV count to:
  * `CatchpointFileHeader`
  * `CatchpointGenerationEvent`
* Catchpoint status (reads) -  Add total, processed, and verified KV counts to:
  * `/v2/status`
  * `goal node status` 

Prior to the PR, _account_ and _KV_ hashes are _not_ disambiguated.  Consequently, it's possible to see catchpoint accounts verified > total accounts like the example below:
```
❯ goal node status -w 1000 -d b4
Last committed block: 0
Sync Time: 4.4s
Catchpoint: 21973000#PJIVBPJMRDKP65YUNCUWJ6WO2ZQ4EKOLFG4S3LA3INYTY4WBCWBQ
Catchpoint total accounts: 24040
Catchpoint accounts processed: 24040
Catchpoint accounts verified: 24207    # non-sensical:  verified > total accounts
...
```

### CODE REVIEW

File: catchup/catchpointService.go [hunk 1]
```diff
@@ -335,12 +339,27 @@ func (cs *CatchpointCatchupService) processStageLedgerDownload() (err error) {
...
+	addedTrieAccountHashes := uint64(0)
+	addedTrieKVHashes := uint64(0)
+	for _, hash := range hashes {
+		if (hash[4] == 3) { // KV
+			addedTrieKVHashes++
+		} else {
+			addedTrieAccountHashes++
+		}
+	}
+
```
Comment (jannotti) [inline catchup/catchpointService.go:356]:
```suggestion
// Estimate VerfiedAccounts based on the portion of account related hashes (account hashes + resource hashes) we have seen.
```

File: catchup/catchpointService.go [hunk 2]
```diff
@@ -335,12 +339,27 @@ func (cs *CatchpointCatchupService) processStageLedgerDownload() (err error) {
...
 	if cs.stats.TotalAccountHashes > 0 {
-		cs.stats.VerifiedAccounts = cs.stats.TotalAccounts * addedTrieHashes / cs.stats.TotalAccountHashes
+		cs.stats.VerifiedAccounts = cs.stats.TotalAccounts * addedTrieAccountHashes / cs.stats.TotalAccountHashes
+	}
+
+	if cs.stats.TotalKVs > 0 { // TODO Is TotalKVHashes needed?
```
Comment (jannotti) [inline catchup/catchpointService.go:361]: I do not think it is. I think it was only needed because of the way we had to estimate accounts due to the combined hashes.

An alternative is to _also_ remove that estimation mechanism, and use the same hash type check to report the actual number of _account_ hashes.

File: ledger/catchpointtracker.go [hunk 3]
```diff
@@ -1081,7 +1084,7 @@ func (ct *catchpointTracker) IsWritingCatchpointDataFile() bool {
 }
 
 // Generates a (first stage) catchpoint data file.
-func (ct *catchpointTracker) generateCatchpointData(ctx context.Context, accountsRound basics.Round, updatingBalancesDuration time.Duration) (uint64 /*totalAccounts*/, uint64 /*totalChunks*/, uint64 /*biggestChunkLen*/, error) {
+func (ct *catchpointTracker) generateCatchpointData(ctx context.Context, accountsRound basics.Round, updatingBalancesDuration time.Duration) (uint64 /*totalKVs*/, uint64 /*totalAccounts*/, uint64 /*totalChunks*/, uint64 /*biggestChunkLen*/, error) {
```
Comment (jannotti) [inline ledger/catchpointtracker.go:1087]: I know you didn't write this, but I don't see why we shouldn't used named return values instead of these comments.

I dislike _naked returns_ that are sometimes used along with named return values. But the named return values themselves serve as useful documentation when the types are the same.

File: ledger/catchupaccessor.go [hunk 4]
```diff
@@ -690,18 +691,22 @@ func (c *catchpointCatchupAccessorImpl) BuildMerkleTrie(ctx context.Context, pro
...
 				}
 				trie.SetCommitter(mc)
-				for _, accountHash := range hashesToWrite {
+				for _, hash := range hashesToWrite {
 					var added bool
-					added, err = trie.Add(accountHash)
+					added, err = trie.Add(hash)
 					if !added {
-						return fmt.Errorf("CatchpointCatchupAccessorImpl::BuildMerkleTrie: The provided catchpoint file contained the same account more than once. hash '%s'", hex.EncodeToString(accountHash))
+						return fmt.Errorf("CatchpointCatchupAccessorImpl::BuildMerkleTrie: The provided catchpoint file contained the same account more than once. hash '%s'", hex.EncodeToString(hash))
```
Comment (jannotti) [inline ledger/catchupaccessor.go:698]: Since this thing is all about debugging, let's improve it - how about telling us what kind of hash it is, by decoding`hash[4]`? We certainly all forgot that fact when debugging - we we presented with a hash that was a problem and didn't immediately look at it and say "Ahh, that's a kv hash".

File: ledger/catchupaccessor.go [hunk 5]
```diff
@@ -690,18 +691,22 @@ func (c *catchpointCatchupAccessorImpl) BuildMerkleTrie(ctx context.Context, pro
...
 						return
 					}
+
 				}
 				uncommitedHashesCount += len(hashesToWrite)
-				hashesWritten += uint64(len(hashesToWrite))
+				if progressUpdates != nil {
+					progressUpdates(hashesToWrite)
+				}
+
```
Comment (jannotti) [inline ledger/catchupaccessor.go:709]: Why the change to call the callback more often?

**Review by jannotti — COMMENTED** [review]
I think this is the right approach.  I suspect it would be possible to get the number of kvs for the catchpoint from the iterator rather that the `select count`, but if it takes any effort to restructure, this seems fine.

File: ledger/catchupaccessor.go (hunk 5 above)
Comment (michaeldiamant) [reply ledger/catchupaccessor.go:709]: @jannotti I'm open to feedback here - I thought it'd be least error prone to:
* Consolidate the hash identification logic in the listener rather than inline here.
* Consequently, the listener receives `[][]byte` rather than `uint64`.  And so I felt it'd be most robust to _avoid_ exposing `hashesWritten` beyond the immediate for-loop.

File: ledger/catchpointtracker.go (hunk 3 above)
Comment (michaeldiamant) [reply ledger/catchpointtracker.go:1087]: @jannotti I have a similar view on naked returns.  Unless you prefer the thought below, I'll add the naked return with named return values.

I had considered returning a partially populated `catchpointFirstStageInfo` struct too because it's created in the subsequent processing step `recordFirstStageInfo`.  I held off because partially populated structs feel error prone to me, but I think it'd be considered a more Go-like approach.

Comment (jannotti) [reply ledger/catchpointtracker.go:1087]: > I'll add the naked return with named return values.

I'm saying you can use named return values, and yet _not_ use naked returns.  This is fine:
```
func two(a, b int) (x int, y int) {
    return a+b, a-b
}
```

File: ledger/catchupaccessor.go (hunk 5 above)
Comment (jannotti) [reply ledger/catchupaccessor.go:709]: I see, yes, makes sense.

File: ledger/catchpointtracker.go (hunk 3 above)
Comment (michaeldiamant) [reply ledger/catchpointtracker.go]: @jannotti I attempted to bring in the feedback via https://github.com/algorand/go-algorand/pull/4808/commits/de8d3d78cfbba3a581e106926cbff7ffb9b93713#diff-9ca6833a989d244ce6b95aaddf4f2f0889203f7e3e6918b0c9ace37c62f21320R1087.  Can you have a look to see if it looks good to you?

File: ledger/catchupaccessor.go (hunk 4 above)
Comment (michaeldiamant) [reply ledger/catchupaccessor.go]: @jannotti Makes sense to me - https://github.com/algorand/go-algorand/pull/4808/commits/aeb4f191aad9c626e10f0c304feea98c0688b0fa#diff-da4e3bb740a328186b93cf1e154745e4f3bdf408866ae584c659b1408db689a9R698 attempts to enrich the error message.  Can you resolve if it matches your expectations?

File: catchup/catchpointService.go (hunk 1 above)
Comment (michaeldiamant) [reply catchup/catchpointService.go:356]: Resolving - I brought in a similar (at least in my mind) comment.  Feel welcomed to make additional suggestions.

**Comment by michaeldiamant** [comment]
As of https://github.com/algorand/go-algorand/pull/4808/commits/85e25e92585d1a0943c5939e81be7dae1e8d1ef4, applying a catchpoint generated by a node running the PR yields the processed == verified counts:

```
❯ goal node status -w 1000 -d b5
Last committed block: 0
Sync Time: 4.9s
Catchpoint: 21997000#4AR2347S3M6CFRVZVFML4ZCIZWMWWZQTZQ5C3AQUTACM573B7HDQ
Catchpoint total accounts: 24061
Catchpoint accounts processed: 24061
Catchpoint accounts verified: 24061
Catchpoint total KVs: 358
Catchpoint KVs processed: 358
Catchpoint KVs verified: 358
Catchpoint total blocks: 1321
Catchpoint downloaded blocks: 32
Genesis ID: betanet-v1.0
Genesis hash: mFgazF+2uRS1tMiL9dsj01hJGySEmPN28B/TjjvpVW0=
```

File: catchup/catchpointService.go [hunk 6]
```diff
@@ -335,12 +339,17 @@ func (cs *CatchpointCatchupService) processStageLedgerDownload() (err error) {
 	return nil
 }
 
-// updateVerifiedAccounts update the user's statistics for the given verified accounts
-func (cs *CatchpointCatchupService) updateVerifiedAccounts(addedTrieHashes uint64) {
+// updateVerifiedCounts update the user's statistics for the given verified hashes
+func (cs *CatchpointCatchupService) updateVerifiedCounts(kvCount, accountCount uint64) {
```
Comment (jannotti) [inline catchup/catchpointService.go:343]: nit: Swap the args? Since we always do these things in the order: account stuff, then kv stuff.

**Review by jannotti — COMMENTED** [review]
I don't feel strongly about swapping the order of kv, accounts all over the place, so leaving an approval.

Earlier we talked about not exporting some stuff, and the KV had to remain exported for some reason. With that in mind, I'd probably add something to those names, as that means `Account` is now an exported symbol.  Seems pretty weird to me.  Maybe AccountHashKind, etc?

I hate going overboard on long names - but these are weird little things that don't feel like they deserve such important names.

**Review by jannotti — APPROVED** [review]

File: catchup/catchpointService.go (hunk 6 above)
Comment (michaeldiamant) [reply catchup/catchpointService.go]: @jannotti I swapped the args in https://github.com/algorand/go-algorand/pull/4808/commits/015199a283e0114df600ac77252b1c25ed1c53c8.  Can you have a look and resolve if it looks good?

File: catchup/catchpointService.go (hunk 2 above)
Comment (michaeldiamant) [reply catchup/catchpointService.go]: Resolving - I've removed the TODO.

**Review by jannotti — DISMISSED** [review]

**Comment by algorandskiy** [comment]
could you make a minimal diff by re-merging master after `initializeHashes` gets it?

File: ledger/catchpointtracker.go [hunk 7]
```diff
@@ -1446,21 +1442,55 @@ func accountHashBuilderV6(addr basics.Address, accountData *baseAccountData, enc
...
+const (
+	accountHK hashKind = iota
+	assetHK
+	appHK
+	kvHK
+)
+
+// HashKindEncodingIndex defines the []byte offset where the hash kind is
+// encoded.
+const HashKindEncodingIndex = 4
```
Comment (algorandskiy) [inline ledger/catchpointtracker.go:1463]: looks not used outside and logically appears to be an implementation detail -> must be not exported

File: ledger/catchpointtracker.go [hunk 8]
```diff
@@ -1446,21 +1442,55 @@ func accountHashBuilderV6(addr basics.Address, accountData *baseAccountData, enc
...
-func resourcesHashBuilderV6(addr basics.Address, cidx basics.CreatableIndex, ctype basics.CreatableType, updateRound uint64, encodedResourceData []byte) []byte {
-	hash := hashBufV6(updateRound, byte(ctype+1)) // one or two ( asset / application ) so we could differentiate the hashes.
+// hashKind enumerates the possible data types hashed into a catchpoint merkle
+// trie.  Each merkle trie hash includes the hashKind byte at a known-offset.
+// By encoding hashKind at a known-offset, it's possible for hash readers to
+// disambiguate the hashed resource.
+//go:generate stringer -type=hashKind
+type hashKind byte
+
+// Defines known kinds of hashes.  Changing an enum ordinal value is a
```
Comment (algorandskiy) [inline ledger/catchpointtracker.go:1452]: double space?

File: ledger/catchpointtracker.go [hunk 9]
```diff
@@ -1446,21 +1442,55 @@ func accountHashBuilderV6(addr basics.Address, accountData *baseAccountData, enc
...
+	assetHK
+	appHK
+	kvHK
+)
+
+// HashKindEncodingIndex defines the []byte offset where the hash kind is
+// encoded.
+const HashKindEncodingIndex = 4
+
+func creatableHashKindFromResourcesData(rd resourcesData, a basics.Address, ci basics.CreatableIndex) (hashKind, error) {
```
Comment (algorandskiy) [inline ledger/catchpointtracker.go:1465]: suggest a shorter name rdGetHashKind(...)

File: ledger/catchpointtracker.go [hunk 10]
```diff
@@ -1446,21 +1442,55 @@ func accountHashBuilderV6(addr basics.Address, accountData *baseAccountData, enc
...
 	binary.LittleEndian.PutUint64(prehash[crypto.DigestSize:], uint64(cidx))
 	copy(prehash[crypto.DigestSize+8:], encodedResourceData[:])
 
-	return finishV6(hash, prehash)
+	return finishV6(hash, prehash), nil
 }
 
 // kvHashBuilderV6 calculates the hash key used for the trie by combining the key and value
 func kvHashBuilderV6(key string, value []byte) []byte {
-	hash := hashBufV6(0, 3) // 3 indicates a kv pair
```
Comment (algorandskiy) [inline ledger/catchpointtracker.go:1463]: `hashBufV6` should use the new constant HashKindEncodingIndex I guess

File: ledger/catchpointtracker.go [hunk 11]
```diff
@@ -1446,21 +1442,55 @@ func accountHashBuilderV6(addr basics.Address, accountData *baseAccountData, enc
...
+	if rd.IsAsset() {
+		return assetHK, nil
+	} else if rd.IsApp() {
+		return appHK, nil
+	}
+	return accountHK, fmt.Errorf("unknown creatable for addr %s, aidx %d, data %v", a.String(), ci, rd)
+}
+
+// resourcesHashBuilderV6 calculates the hash key used for the trie by combining the creatable's resource data and its index
+func resourcesHashBuilderV6(rd resourcesData, addr basics.Address, cidx basics.CreatableIndex, updateRound uint64, encodedResourceData []byte) ([]byte, error) {
```
Comment (algorandskiy) [inline ledger/catchpointtracker.go:1475]: rd resourcesData can be big, pass a pointer? or call creatableHashKindFromResourcesData in-place before resourcesHashBuilderV6

**Review by algorandskiy — COMMENTED** [review]
The change looks good, some code-style comments

File: ledger/catchpointtracker.go (hunk 10 above)
Comment (michaeldiamant) [reply ledger/catchpointtracker.go:1463]: @algorandskiy I'm not following the request - let me know if it's still a pending concern after reading the note below?

The thread highlights _before_ not _after_ the PR.  The corresponding line _is_ changed to use the constant.  So, I don't _think_ there's further action here.

File: ledger/catchpointtracker.go (hunk 7 above)
Comment (michaeldiamant) [reply ledger/catchpointtracker.go]: @algorandskiy Thanks for catching - I previously needed it to be exposed, but that's no longer the case.  Fixed in https://github.com/algorand/go-algorand/pull/4808/commits/351f9bec42aa591d4fbad69957924f2ab51a49f9.  Can you resolve if it looks good?

File: ledger/catchpointtracker.go (hunk 8 above)
Comment (michaeldiamant) [reply ledger/catchpointtracker.go]: @algorandskiy I have an old habit of double spacing after periods and colons.  😅 I changed it in https://github.com/algorand/go-algorand/pull/4808/commits/c9a8b11d2a15f3c7ea2fdc169b247c083b888b2b.  I'll leave thread open for you to resolve.

File: ledger/catchpointtracker.go (hunk 9 above)
Comment (michaeldiamant) [reply ledger/catchpointtracker.go]: @algorandskiy I refactored to a shorter name in https://github.com/algorand/go-algorand/pull/4808/commits/5fd57700d86ed8e23dc6b97ed0f9926559e86a5f.  Can you resolve if it looks good?

File: ledger/catchpointtracker.go (hunk 11 above)
Comment (michaeldiamant) [reply ledger/catchpointtracker.go]: @algorandskiy Good call out - Sorry I missed it.  I changed to pointer in https://github.com/algorand/go-algorand/pull/4808/commits/56b4a54868b913bfec8ec6a9803e2d156dafebf8 - Can you resolve if it looks good?

File: ledger/catchpointtracker.go (hunk 10 above)
Comment (algorandskiy) [reply ledger/catchpointtracker.go:1463]: >	hash[hashKindEncodingIndex] = byte(kind)
well, I overlooked it somehow. apologies.

**Review by algorandskiy — APPROVED** [review]
